confgenerator/confgenerator.go

```diff
@@ -96,55 +94,272 @@ func (uc *UnifiedConfig) GenerateOtelConfig(hostInfo *host.InfoStat) (string, er
+						metricNames = append(metricNames, fmt.Sprintf(`^%s$`, strings.Join(literals, `.*`)))
+					}
+					excludeMetrics := otel.ExcludeMetrics{
+						ExcludeMetricsID: "filter/exclude_" + processorID,
+						MetricNames:      metricNames,
+					}
+					processorList = append(processorList, &excludeMetrics)
+					processorMap[processorID] = &excludeMetrics
+				}
+			}
+		}
+	}
+	return processorList, processorMap, nil
+}
+
+func generateOtelServices(receiverMap map[string]otel.Receiver, exporterMap map[string]otel.Exporter, processorMap map[string]otel.Processor, pipelines map[string]*MetricsPipeline) ([]*otel.Service, error) {
 	serviceList := []*otel.Service{}
 	for _, pID := range sortedKeys(pipelines) {
 		p := pipelines[pID]
 		for _, rID := range p.ReceiverIDs {
-			var pipelineID string
-			var defaultProcessors []string
-			if strings.HasPrefix(receiverNameMap[rID], "hostmetrics/") {
-				defaultProcessors = []string{"agentmetrics/system", "filter/system", "metricstransform/system", "resourcedetection"}
-				pipelineID = "system"
-			} else if strings.HasPrefix(receiverNameMap[rID], "windowsperfcounters/mssql") {
-				defaultProcessors = []string{"metricstransform/mssql", "resourcedetection"}
-				pipelineID = "mssql"
-			} else if strings.HasPrefix(receiverNameMap[rID], "windowsperfcounters/iis") {
-				defaultProcessors = []string{"metricstransform/iis", "resourcedetection"}
-				pipelineID = "iis"
+			r, ok := receiverMap[rID]
+			if !ok {
+				panic(fmt.Sprintf("Internal error: receiver %q not found", rID))
 			}
 
 			var processorIDs []string
-			processorIDs = append(processorIDs, defaultProcessors...)
+			processorIDs = append(processorIDs, r.DefaultProcessors()...)
 			for _, processorID := range p.ProcessorIDs {
-				processorIDs = append(processorIDs, processorNameMap[processorID])
+				processorIDs = append(processorIDs, processorMap[processorID].GetID())
 			}
 
 			var pExportIDs []string
 			for _, eID := range p.ExporterIDs {
-				pExportIDs = append(pExportIDs, exporterNameMap[eID])
+				pExportIDs = append(pExportIDs, exporterMap[eID].GetID())
 			}
 			service := otel.Service{
-				ID:         pipelineID,
-				Receivers:  fmt.Sprintf("[%s]", receiverNameMap[rID]),
-				Processors: fmt.Sprintf("[%s]", strings.Join(processorIDs, ",")),
-				Exporters:  fmt.Sprintf("[%s]", strings.Join(pExportIDs, ",")),
+				ID:         r.DefaultPipelineID(),
+				Receivers:  []string{r.GetID()},
+				Processors: processorIDs,
+				Exporters:  pExportIDs,
 			}
 			serviceList = append(serviceList, &service)
 		}
 	}
 	return serviceList, nil
 }
 
+// GenerateFluentBitConfigs generates FluentBit configuration from unified agents configuration
+// in yaml. GenerateFluentBitConfigs returns empty configurations without an error if `logging`
+// does not exist as a top-level field in the input yaml format.
+func (uc *UnifiedConfig) GenerateFluentBitConfigs(logsDir string, stateDir string, hostInfo *host.InfoStat) (string, string, error) {
```

Comment:
This is an awful lot of repeated boilerplate code. Why does this need a separate struct for each of these? Can't we emit a single struct, or at least one struct for filters, another for inputs, etc.?

Reply:
Getting there.